src/server/controllers/VerificationController.ts

```diff
@@ -23,48 +27,95 @@ export default class VerificationController extends BaseController implements IC
+        const addresses = this.validateAddresses(req.body.address);
+        return { chain, addresses };
+    }
+
+    private legacyVerifyEndpoint = async (req: Request, res: Response): Promise<any> => {
+        const { addresses, chain } = this.validateChainAndAddresses(req);
+
+        for (const address of addresses) {
+            const result = await this.verificationService.findByAddress(address, chain, config.repository.path);
+            if (result.length != 0) {
+                return res.status(200).send({ result });
             }
-            inputData.contracts = validatedContracts;
-            const matches: any = [];
-            matches.push(await this.verificationService.inject(inputData, config.localchain.url));
-            Promise.all(matches).then((result) => {
-                res.status(200).send({ result })
-            }).catch()
         }
 
+        const inputFiles = this.extractFiles(req); // throws
+        if (!inputFiles) {
+            const msg = "The contract at the provided address has not yet been sourcified.";
+            throw new NotFoundError(msg);
+        }
+
+        let validatedContracts: CheckedContract[];
+        try {
+            validatedContracts = this.validationService.checkFiles(inputFiles);
+        } catch(error) {
+            throw new BadRequestError(error.message);
+        }
+
+        const errors = validatedContracts
+                        .filter(contract => Object.keys(contract.invalid).length)
+                        .map(contract => `${contract.name} ${Object(contract.invalid).keys()}`);
+        if (errors.length) {
+            throw new BadRequestError("Errors in:\n" + errors.join("\n"), false);
+        }
+
+        if (validatedContracts.length !== 1) {
+            const contractNames = validatedContracts.map(c => c.name).join(", ");
+            const msg = `Detected ${validatedContracts.length} contracts (${contractNames}), but can only verify 1 at a time.`;
+            throw new BadRequestError(msg);
+        }
+
+        const contract = validatedContracts[0];
+        if (!contract.compilerVersion) {
+            throw new BadRequestError("Metadata file not specifying a compiler version.");
+        }
+        const inputData: InputData = { contract, addresses, chain };
+
+        const resultPromise = this.verificationService.inject(inputData, config.localchain.url);
+        resultPromise.then(result => {
+            res.status(200).send({ result: [result] }); // TODO array is an old expected behavior (e.g. by frontend)
+        }).catch(error => {
+            res.status(500).send({ error: error.message });
+        });
     }
 
-    checkByAddresses = async (req: any, res: Response) => {
-        let resultArray: Array<Object> = [];
+    private checkByAddresses = async (req: any, res: Response) => {
+        const missingParams = ["addresses", "chainIds"].filter(p => !req.query[p]);
```

Comment:
We can probably use the express-validator here to have more consistent code.
Example is in the FileController.ts